### PR TITLE
Expose Default values in Function definition to AST

### PR DESCRIPTION
**Pull request by Thirumalai-Shaktivel**

### CODE REVIEW

**Comment by Thirumalai-Shaktivel** [comment]
ping @akshanshbhatt

**Comment by certik** [comment]
I would need to review this one more carefully, it seems a bit complex, but probably we have to do it this way?

**Comment by akshanshbhatt** [comment]
> ping @akshanshbhatt

Awesome! I tried it locally, and it seems to work as intended.
Any idea why the MacOS CI is failing?

**Comment by certik** [comment]
I tested locally on macOS, the test passes. At the CI the failure seems to be
```
TEST: parser/function_def1.py
    * ast_new The JSON metadata differs against reference results
Reference JSON: tests/reference/ast_new-function_def1-1a872df.json
Output JSON:    tests/output/ast_new-function_def1-1a872df.json
Omitting 8 identical items
Differing items:
{'returncode': -11} != {'returncode': 0}
{'stdout': None} != {'stdout': 'ast_new-function_def1-1a872df.stdout'}
{'stdout_hash': None} != {'stdout_hash': 'ad5ea65f2ef1994819a5c48936f936ea955303c08253419d72aa450b'}
```
Which is one of the tests modified in this PR, it seems at the CI LPython failed with a returncode "-1" and did not produce any output nor errors.

This could happen if lpython segfaulted, possibly if some pointer handling in this PR is incorrect and just happened to be exposed at the CI.

File: src/lpython/parser/semantics.h [hunk 1]
```diff
@@ -283,76 +284,131 @@ static inline withitem_t *WITH_ITEM(Allocator &al, Location &l,
...
-        expr_t** m_defaults, size_t n_defaults) {
+        Arg* /*m_posonlyargs*/, size_t /*n_posonlyargs*/,
+        Arg** m_args, size_t n_args,
+        Arg* m_vararg, size_t n_vararg,
+        Arg** m_kwonlyargs, size_t n_kwonlyargs,
+        expr_t** /*m_kw_defaults*/, size_t /*n_kw_defaults*/,
+        Arg* m_kwarg, size_t n_kwarg) {
     Args *r = al.allocate<Args>();
+    Vec<arg_t> _m_args;
+    _m_args.from_pointer_n(nullptr, 0);
```
Comment (certik) [inline src/lpython/parser/semantics.h:310]: I don't think you need this line?

File: src/lpython/parser/semantics.h [hunk 2]
```diff
@@ -283,76 +284,131 @@ static inline withitem_t *WITH_ITEM(Allocator &al, Location &l,
...
+        _m_vararg.reserve(al, n_vararg);
+        defaults.reserve(al, n_vararg);
+        _m_vararg.push_back(al, m_vararg->_arg);
+        if(m_vararg->def) {
+            defaults.push_back(al, m_vararg->defaults);
+        }
+    }
+    if(n_kwonlyargs > 0) {
+        _m_kwonlyargs.reserve(al, n_kwonlyargs);
+        defaults.reserve(al, n_kwonlyargs);
```
Comment (certik) [inline src/lpython/parser/semantics.h:339]: You seem to be calling "reserve" couple times on `defaults`. That might be the issue.

Comment (certik) [reply src/lpython/parser/semantics.h:339]: Replace this:
```

    Vec<arg_t> _m_args;
    _m_args.from_pointer_n(nullptr, 0);
```
with:
```
    Vec<arg_t> _m_args;
    _m_args.reserve(al, 4);
```
The same for all the other ones. Remove all calls to reserve beyond that. I don't know if this fixes the issue at the CI, but that's one issue I can see.

**Comment by Thirumalai-Shaktivel** [comment]
> I would need to review this one more carefully, it seems a bit complex, but probably we have to do it this way?

Yup, I don't think we have other option!

File: src/lpython/parser/semantics.h (hunk 2 above)
Comment (Thirumalai-Shaktivel) [reply src/lpython/parser/semantics.h:339]: Yup, I will do it.

**Comment by certik** [comment]
One possible issue is eliminated. But the bug is still there.

You can now try to split this PR into smaller pieces and "bisect" it. You can open a separate PR for the debugging.

**Comment by Thirumalai-Shaktivel** [comment]
This PR is ready to go in!

**Review by certik — APPROVED** [review]

**Pull request merged**